### DISCUSSION TITLE
FFRPG: Japanese characters in the compilation list are garbled

### DISCUSSION THREAD

**Discussion by koinumasaoko** [codefori/vscode-ibmi#1783] [Q&A] [answered]
Hello,
This question is about something happening in the Japanese environment.
Currently, I am testing the development of FFRPG using VSCode（Code for IBMi）.
In the version 7.5 environment, only Japanese characters in the compile list are garbled.
Could you please give me some advice regarding the garbled characters?
This does not occur in version 7.4, so I would like to clarify the cause.

■Verification procedure

1. Connect to IBMi server with VSCode

2. Right-click the source in IFS BROWSER and select Run Action to compile.

3. Click Open Output to see the compilation list. However, the Japanese characters become garbled.

- The compile command is as follows.
`CRTBNDRPG PGM(QGPL/icafe016) SRCSTMF('/home/KOINUMA/ffrpgsrc/icafe016.rpgle') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTRLS(*CURRENT) TGTCCSID(*JOB)`
![image](https://github.com/codefori/vscode-ibmi/assets/56465596/ef89c86b-0a8f-4d56-8487-9549100b1eb9)

■Environmental assumptions
- primary language：2930
- Version：7.5
- QCCSID：1399 
- IBM PASE for i environment variables： QIBM_PASE_CCSID=1208
- CCSID of user profile：SYSVAL
- FFRPG source encoding: UTF8

■Other concerns
I set up to automatically put the source from the PC to IFS using SFTP.
In version 7.5, the STMF CCSID is  ~~817~~  **819,（The ccsid was written incorrectly. sorry…）**
In version 7.4, the STMF CCSID is 1208, which is different.

**Comment by chrjorgensen** [comment]
@koinumasaoko

Hello,

If your IBM i 7.4 system works and your IBM i 7.5 system does not work, you should check your system settings between the systems.
Also check that your two connections in Code for IBM i have identical actions setup.
Your garbled characters is most likely a CCSID issue on your 7.5 system.

Try connecting to your systems in a SSH client and run the following commands:

```
qsh (note, there will be no output from this command)
system "CRTBNDRPG PGM(QGPL/icafe016) SRCSTMF('/home/KOINUMA/ffrpgsrc/icafe016.rpgle') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTRLS(*CURRENT) TGTCCSID(*JOB)"
(Press Ctrl-D to exit qsh)
```

If the output from your 7.5 system is still garbled, we can eliminate Code for IBM i as the cause...

**Reply by sebjulliand** [reply]
@chrjorgensen , for your information, there was a thread started here: https://community.ibm.com/community/user/power/discussion/visual-studio-codecode-for-ibmiffrpg
There are some interesting information, like the uploaded files CCSID difference between 7.4 and 7.5.
@koinumasaoko , the point raised on the Tech Xchange thread may be a good lead. You raised the fact that transferring your files using SFTP created files with CCSID 1208 (Unicode) in 7.4 and 817 (ISO-8859-1) in 7.5. That would be good to check why this happens (i.e. check the system configuration). It sounds like the output generated by the compilation is affected by this too.
You can also try to open an SSH session using a external tool (like PUTTY) and run the compilation command to see if the output is garbled as well, outside of VSCode.

**Reply by chrjorgensen** [reply]
@sebjulliand Yeah, saw the post in the IBM community before coming here... 😉 

I agree that the different CCSID of the streamfiles is an issue - but the output from the RPG compiler is also garbled, even with text coming from the compiler and not from the source (streamfile). So I think there is a general CCSID issue with the PASE environment on the 7.5 system...

**Reply by koinumasaoko** [reply]
@chrjorgensen @sebjulliand 
Thank you for your comment. STMF was 819 instead of 817. I'm very sorry for the typo.
Now, by changing the system environment variable, I was able to force the SMTF CCSID to 1208.
`ADDENVVAR ENVVAR(PASE_DEFAULT_UTF8) VALUE(Y) LEVEL(*SYS)`
***
> Try connecting to your systems in a SSH client and run the following commands:

> You can also try to open an SSH session using a external tool (like PUTTY) and run the compilation command to see if the output is garbled as well, outside of VSCode.

Thank you for your advice.
I checked these (using PuTTY).
Still, the compilation result is garbled.

**Reply by koinumasaoko** [reply]
`export LANG=ja_JP.UTF-8`

When using PuTTY with the above settings, garbled characters did not occur!

**Reply by chrjorgensen** [reply]
@koinumasaoko The environment variable `PASE_DEFAULT_UTF8` should not be necessary on a correctly configured IBM i 7.5 system - I don't have it on my 7.5 system and PASE is still all UTF-8 here. I don't have any `QIBM_PASE_CCSID` either.

This is definitely a language setup / system configuration issue and not a Code for IBM i issue, but anyway I would suggest you do the following:

- check the system values QCCSID, QCNTRYID and QLANGID are correct for Japanese. I run Danish (EBCDIC 277), so my system has QCCSID = 277, QCNTRYID = 'DK' and QLANGID = 'DAN'. The system values should cover the setup used by most users on the system.
- check your user profile uses the system values, i.e. has *SYSVAL in LANGID, CNTRYID and CCSID.
- check your system-wide environment variables (*SYS). I have no system-wide or environment variables defined on my 7.5 system.
- check you are running the SSH server from licensed program 5733-SC1 (IBM Portable Utilities for i). We've seen people running the open source version of OpenSSH (from IBM YUM repositories) instead and having all kinds of issues. You can check it several ways, but one way to do it is by running the command `/QOpenSys/usr/bin/ps -eo pid,args` in a SSH terminal (PuTTY). You should get a line showing `/QOpenSys/QIBM/ProdData/SC1/OpenSSH/sbin/sshd -R`, which is the the IBM delivered SSH server (in `/QIBM/ProdData`).
- as always, check you are up to date on the PTF's (may not be relevant for the encoding issue, but definitely is relevant for the security protocols used by SSH).

After you have corrected the errors in your language configuration, you should restart the SSH server (`ENDTCPSVR *SSHD` and then `STRTCPSVR *SSHD`) which should now run correctly in UTF-8 (1208). And then you can check if the compilation command now shows correct output in PuTTY.

**Reply by koinumasaoko** [reply]
@chrjorgensen 
Thank you for your support.
I checked the things you told me.

1. SYSVAL
```
QCCSID：1399
QCNTRYID： JP
QLANGID：JPN
```
2. user profile
```
LANGID：*SYSVAL
CNTRYID：*SYSVAL
CCSID：*SYSVAL
```
3. WRKENVVAR
- WRKENVVAR LEVEL(*SYS)
```
 PASE_DEFAULT_UTF8           'Y'
```
- WRKENVVAR LEVEL(*JOB)
```
PASE_DEFAULT_UTF8           'Y'
LANG                        '/QSYS.LIB/JA_JP_1399.LOCALE'
```
4. SSH 
- I confirmed the following with Putty.
```
-bash-5.1$ /QOpenSys/usr/bin/ps -eo pid,args~
  PID COMMAND
/QOpenSys/QIBM/ProdData/SC1/OpenSSH/sbin/sshd -R
```
5. PTF
- I think the PTF is the latest one.

***
When setting environment variables in /home/KOINUMA[user profile]/.profile, garbled characters no longer occur when connecting with SSH using Putty or PASE.
```
export LANG=ja_JP.UTF-8
```

There is one thing that bothers me.
Open the terminal with Terminals → PASE →
Garbled characters do not occur when you run the compile command.
![image](https://github.com/codefori/vscode-ibmi/assets/56465596/ae97afbb-0962-48f4-9bcc-950ab9dbdec5)

```
qsh
system "CRTBNDRPG PGM(QGPL/icafe016) SRCSTMF('/home/KOINUMA/ffrpgsrc/icafe016.rpgle') OPTION(*EVENTF) DBGVIEW(*SOURCE) TGTRLS(*CURRENT) TGTCCSID(5035)"
```
However, when I execute Run Action (Create Bound RPG Program (CRTBNDRPG)),
Both the compile list and the terminal where the compilation result log is output are garbled.

**Reply by chrjorgensen** [reply]
@koinumasaoko 

Why do you have these settings?

```
PASE_DEFAULT_UTF8           'Y'
LANG                        '/QSYS.LIB/JA_JP_1399.LOCALE'
```

And what happens if you run without these settings? Or with only

```
LANG                        '/QSYS.LIB/JA_JP.LOCALE'
```


I would still advice to compare the settings of the IBM i 7.4 system having no problems and the IBM i 7.5 having problems - to my knowledge, there are no functional differences between 7.4 and 7.5, and I think the problems are caused by some differences in the settings.

Be aware that  bash shell setup can be different in interactive logon and command execution. Code for IBM i uses the latter (and does not sign in to any terminal), but your PASE terminal opened in Code for IBM i or PuTTY does. It is described in [more detail here](https://www.gnu.org/software/bash/manual/html_node/Bash-Startup-Files.html), but the essence is that your `.profile` file is not read when there is no interactive login. So be careful to put statements that sets the environment in this file, since it is not always read.

If you still have no luck in eliminating this server error, I would advice to create a support case at IBM Support. They may help you with your settings - and probably have more experience with Japanese than I do. :laughing:

**Reply by koinumasaoko** [reply]
@chrjorgensen 
Thank you for teaching me so much. I would also like to thank you for your polite reply to my poor English.
I will check with my team members about the ENVVAR settings.
I also understood that Code for IBM i's login method is a command execution type, which executes without reading the .profile.
I think I'll try creating a case as well.
It seems like a lot of special things are happening.
I think with your help we are getting closer to solving the problem. thank you very much!

**Reply by chrjorgensen** [reply, marked as answer]
@koinumasaoko

You're welcome - we hope you will get this fixed soon and can enjoy using Code for IBM i to its full extent. :smiley:

**Comment by koinumasaoko** [comment]
@chrjorgensen 
Hello!
We would like to inform you that the problem has been resolved.
This issue was resolved by applying cumulative PTFs and non-cumulative group PTFs.
If you know of any affected PTFs, could you please let me know?

**Reply by chrjorgensen** [reply]
@koinumasaoko Thanks for reporting back on your success!

Best advice I can give is to always keep current on PTF's - install cumulative PTF and group PTF's when they're updated.
If your test/development system have the same release as the production system (it should have!), then first install on test/development system (if available) and when no problems have appeared, then install on production system.
